File: saas_register/saas_register/page/monthly_cost_entry/monthly_cost_entry.py
"""Whitelisted endpoints for the Monthly Cost Entry page.

The page is a Frappe Page (not a doctype) — see monthly_cost_entry.json + .js.
Calls these methods to load/save SaaS Monthly Cost rows by (application, month).

Per-row currency: each SaaS Monthly Cost stores its own `currency` and
`exchange_rate_to_base`, so AWS can bill in USD while Spend Dashboard reports
in AED. The page picker auto-fills exchange_rate from ERPNext's Currency
Exchange table (callers can override).

Audit trail: every upsert writes a row to SaaS Monthly Cost Audit. Both insert
(no old_amount) and update (old → new) paths are recorded with action,
edited_by, edited_at, and source.
"""

from __future__ import annotations

import json
import re
from datetime import date, timedelta
from difflib import SequenceMatcher

import frappe
from frappe import _
from frappe.utils import add_months, flt, get_first_day, getdate, today

from saas_register.saas_register.cost_audit import write_audit


# ---------------------------------------------------------------------------
# GET: grid
# ---------------------------------------------------------------------------


@frappe.whitelist()
def get_grid(month: str, status: str = "Active", category: str | None = None) -> list[dict]:
	"""Return one row per matching SaaS Application with:
	- app, app_name, vendor, subscription_model, cost_center, currency
	- expected_amount (avg of previous 3 months, converted to base currency)
	- actual_amount (this month's row if it exists)
	- actual_currency + exchange_rate_to_base (per-row, defaults from app)
	- notes (this month's row notes)
	"""
	target_month = get_first_day(getdate(month))
	base_currency = _base_currency()

	app_filters: dict = {}
	if status and status != "All":
		app_filters["status"] = status
	if category:
		app_filters["category"] = category

	apps = frappe.get_all(
		"SaaS Application",
		filters=app_filters,
		fields=["name", "app_name", "vendor", "subscription_model", "cost_center", "currency", "monthly_cost"],
		order_by="app_name asc",
	)

	rows: list[dict] = []
	for app in apps:
		previous_months = _months_before(target_month, 3)
		hist = frappe.get_all(
			"SaaS Monthly Cost",
			filters={
				"parent": app["name"],
				"parenttype": "SaaS Application",
				"month": ("in", previous_months),
			},
			fields=["month", "amount", "exchange_rate_to_base"],
		)
		# Expected (avg of last 3) is reported in base currency so the Δ column
		# is meaningful across multi-currency apps.
		if hist:
			converted = [flt(h.amount) * flt(h.exchange_rate_to_base or 1.0) for h in hist]
			expected = sum(converted) / len(converted)
		else:
			expected = flt(app.get("monthly_cost"))

		this_month = frappe.db.get_value(
			"SaaS Monthly Cost",
			{
				"parent": app["name"],
				"parenttype": "SaaS Application",
				"month": target_month,
			},
			["amount", "notes", "currency", "exchange_rate_to_base"],
			as_dict=True,
		)

		row_currency = (this_month and this_month.currency) or app.get("currency") or base_currency
		row_rate = (this_month and flt(this_month.exchange_rate_to_base)) or _suggested_rate(row_currency, base_currency, target_month)

		rows.append(
			{
				"app": app["name"],
				"app_name": app["app_name"],
				"vendor": app["vendor"],
				"subscription_model": app["subscription_model"],
				"cost_center": app["cost_center"],
				"currency": row_currency,
				"exchange_rate_to_base": row_rate,
				"base_currency": base_currency,
				"expected_amount": flt(expected),
				"actual_amount": flt(this_month.amount) if this_month else None,
				"notes": (this_month.notes if this_month else None),
			}
		)

	return rows


def _months_before(month: date, n: int) -> list[date]:
	out: list[date] = []
	for i in range(1, n + 1):
		out.append(add_months(month, -i))
	return out


def _base_currency() -> str:
	# Settings field is still named `default_currency` for back-compat. Spec calls
	# it base_currency; the relabel makes the role clear in the form.
	return frappe.db.get_single_value("SaaS Register Settings", "default_currency") or "AED"


@frappe.whitelist()
def get_exchange_rate(from_currency: str, to_currency: str, on_date: str | None = None) -> float:
	"""Look up the ERPNext Currency Exchange rate, falling back to 1.0 when
	currencies match or no rate is configured."""
	if not from_currency or not to_currency or from_currency == to_currency:
		return 1.0
	return _suggested_rate(from_currency, to_currency, getdate(on_date) if on_date else getdate(today()))


def _suggested_rate(from_currency: str, to_currency: str, on_date) -> float:
	if not from_currency or not to_currency or from_currency == to_currency:
		return 1.0
	# Most-recent Currency Exchange row at or before the target month.
	rate = frappe.db.sql(
		"""
		SELECT exchange_rate
		FROM `tabCurrency Exchange`
		WHERE from_currency = %s AND to_currency = %s AND date <= %s
		ORDER BY date DESC
		LIMIT 1
		""",
		(from_currency, to_currency, on_date),
	)
	if rate and rate[0][0]:
		return flt(rate[0][0])
	return 1.0


# ---------------------------------------------------------------------------
# UPSERT: one cell
# ---------------------------------------------------------------------------


@frappe.whitelist()
def upsert_monthly_cost(
	application: str,
	month: str,
	amount: float | str,
	currency: str | None = None,
	exchange_rate_to_base: float | str | None = None,
	notes: str | None = None,
) -> dict:
	"""Upsert a SaaS Monthly Cost row identified by (parent=application, month).

	Writes a SaaS Monthly Cost Audit row in the same transaction.
	"""
	target_month = get_first_day(getdate(month))
	new_amount = flt(amount)

	app = frappe.get_doc("SaaS Application", application)
	frappe.has_permission("SaaS Application", "write", app, throw=True)

	# Default currency from the app; exchange rate from Currency Exchange.
	new_currency = currency or app.currency or _base_currency()
	if exchange_rate_to_base in (None, "", 0):
		new_rate = _suggested_rate(new_currency, _base_currency(), target_month)
	else:
		new_rate = flt(exchange_rate_to_base) or 1.0

	existing_row = None
	for row in (app.monthly_costs or []):
		if getdate(row.month) == target_month:
			existing_row = row
			break

	old_amount = flt(existing_row.amount) if existing_row else None
	old_currency = existing_row.currency if existing_row else None
	action = "Update" if existing_row else "Insert"

	if existing_row:
		existing_row.amount = new_amount
		existing_row.currency = new_currency
		existing_row.exchange_rate_to_base = new_rate
		if notes is not None:
			existing_row.notes = notes
		existing_row.source = "Manual"
		existing_row.last_edited_by = frappe.session.user
		existing_row.last_edited_at = frappe.utils.now()
	else:
		app.append(
			"monthly_costs",
			{
				"month": target_month,
				"amount": new_amount,
				"currency": new_currency,
				"exchange_rate_to_base": new_rate,
				"source": "Manual",
				"notes": notes or "",
				"last_edited_by": frappe.session.user,
				"last_edited_at": frappe.utils.now(),
			},
		)

	# Skip the parent-diff auto-audit — we'll write the audit ourselves below
	# with the right action label and full context.
	app.flags.skip_cost_audit = True
	app.save(ignore_permissions=False)

	write_audit(
		saas_application=application,
		month=target_month,
		action=action,
		old_amount=old_amount,
		old_currency=old_currency,
		new_amount=new_amount,
		new_currency=new_currency,
		source="Manual",
	)

	return {
		"application": application,
		"month": str(target_month),
		"amount": new_amount,
		"currency": new_currency,
		"exchange_rate_to_base": new_rate,
	}


# ---------------------------------------------------------------------------
# Paste from spreadsheet
# ---------------------------------------------------------------------------


@frappe.whitelist()
def paste_monthly_costs(month: str, rows: str | list, commit: int = 0) -> dict:
	"""Fuzzy-match each row against SaaS Application.app_name. Returns
	{matched, unmatched}. When commit=1, also upserts the matched rows.

	Optional `currency` per row honours per-row currency; otherwise the app's
	currency is used and the exchange rate is fetched from Currency Exchange.
	"""
	target_month = get_first_day(getdate(month))
	parsed = _ensure_rows(rows)
	apps = frappe.get_all("SaaS Application", fields=["name", "app_name"])

	matched: list[dict] = []
	unmatched: list[dict] = []

	for r in parsed:
		input_name = (r.get("app_name") or "").strip()
		amount = flt(r.get("amount"))
		currency = (r.get("currency") or "").strip() or None
		if not input_name:
			continue

		app = _fuzzy_match(input_name, apps)
		if app:
			matched.append(
				{
					"input": input_name,
					"app": app["name"],
					"app_name": app["app_name"],
					"amount": amount,
					"currency": currency,
				}
			)
			if int(commit):
				upsert_monthly_cost(
					application=app["name"],
					month=str(target_month),
					amount=amount,
					currency=currency,
				)
		else:
			unmatched.append({"app_name": input_name, "amount": amount, "currency": currency})

	return {"matched": matched, "unmatched": unmatched}


def _ensure_rows(rows):
	if isinstance(rows, str):
		return json.loads(rows)
	return rows or []


_NORMALIZE_RE = re.compile(r"[^a-z0-9]+")


def _normalize(s: str) -> str:
	return _NORMALIZE_RE.sub("", (s or "").lower())


def _fuzzy_match(input_name: str, apps: list[dict]) -> dict | None:
	"""Return the SaaS Application dict that best matches input_name. None if
	no app crosses the similarity threshold."""
	if not input_name:
		return None
	target = _normalize(input_name)

	best: tuple[float, dict | None] = (0.0, None)
	for app in apps:
		candidate = _normalize(app["app_name"])
		# Substring match → high confidence
		if candidate and candidate in target:
			score = 0.95
		elif target and target in candidate:
			score = 0.95
		else:
			score = SequenceMatcher(None, target, candidate).ratio()
		if score > best[0]:
			best = (score, app)

	if best[0] >= 0.8:
		return best[1]
	return None

File: saas_register/saas_register/page/monthly_cost_entry/monthly_cost_entry.js
// Monthly Cost Entry page
// Spreadsheet-style grid for Finance to capture actual monthly SaaS spend.

frappe.pages["monthly-cost-entry"].on_page_load = function (wrapper) {
	const page = frappe.ui.make_app_page({
		parent: wrapper,
		title: __("Monthly Cost Entry"),
		single_column: true,
	});

	new MonthlyCostEntry(page);
};

class MonthlyCostEntry {
	constructor(page) {
		this.page = page;
		this.rows = [];
		this.dirty = new Set();
		this.month = frappe.datetime.month_start();
		this.status_filter = "Active";
		this.category_filter = "";
		this.search_q = "";

		this.setup_toolbar();
		this.render_skeleton();
		this.load();
	}

	// ---------------------------------------------------------------------
	// Toolbar
	// ---------------------------------------------------------------------

	setup_toolbar() {
		const me = this;

		this.page.add_field({
			fieldtype: "Date",
			fieldname: "month",
			label: __("Month"),
			default: this.month,
			change: () => {
				me.month = frappe.datetime.month_start(me.page.fields_dict.month.get_value());
				me.load();
			},
		});

		this.page.add_field({
			fieldtype: "Select",
			fieldname: "status",
			label: __("App Status"),
			options: "Active\nReview\nKill-Replace\nAll",
			default: "Active",
			change: () => {
				me.status_filter = me.page.fields_dict.status.get_value() || "Active";
				me.load();
			},
		});

		this.page.add_field({
			fieldtype: "Link",
			fieldname: "category",
			label: __("Category"),
			options: "SaaS Category",
			change: () => {
				me.category_filter = me.page.fields_dict.category.get_value() || "";
				me.load();
			},
		});

		this.page.add_field({
			fieldtype: "Data",
			fieldname: "search",
			label: __("Search"),
			change: () => {
				me.search_q = (me.page.fields_dict.search.get_value() || "").toLowerCase();
				me.render_rows();
			},
		});

		this.page.set_primary_action(__("Save All"), () => me.save_all());
		this.page.add_menu_item(__("Paste from spreadsheet"), () => me.show_paste_dialog());
	}

	render_skeleton() {
		this.page.main.html(`
			<div class="mce-wrap" style="padding: 12px 0;">
				<div class="mce-status" style="margin-bottom:8px;color:var(--text-muted);font-size:12px"></div>
				<table class="table table-bordered mce-grid" style="font-size:13px;">
					<thead style="background:#FAFBFC">
						<tr>
							<th style="width:28%">Application</th>
							<th>Model</th>
							<th>Cost Center</th>
							<th class="text-right">Expected</th>
							<th class="text-right" style="background:#FFFBEA">Actual (${frappe.datetime.global_date_format(this.month)})</th>
							<th class="text-right">Δ%</th>
							<th>Note</th>
						</tr>
					</thead>
					<tbody class="mce-body">
						<tr><td colspan="7" class="text-center text-muted">${__("Loading…")}</td></tr>
					</tbody>
				</table>
			</div>
		`);
	}

	// ---------------------------------------------------------------------
	// Data
	// ---------------------------------------------------------------------

	async load() {
		const me = this;
		this.page.main.find(".mce-body").html(
			`<tr><td colspan="7" class="text-center text-muted">${__("Loading…")}</td></tr>`
		);
		const r = await frappe.call({
			method: "saas_register.saas_register.page.monthly_cost_entry.monthly_cost_entry.get_grid",
			args: {
				month: this.month,
				status: this.status_filter,
				category: this.category_filter,
			},
		});
		this.rows = r.message || [];
		this.dirty.clear();
		this.update_header();
		this.render_rows();
	}

	update_header() {
		this.page.main.find(".mce-status").text(
			__("{0} apps · month {1}", [this.rows.length, frappe.datetime.global_date_format(this.month)])
		);
	}

	render_rows() {
		const me = this;
		const tbody = this.page.main.find(".mce-body");
		const q = me.search_q;
		const rows = q
			? me.rows.filter(
					(r) =>
						(r.app_name || "").toLowerCase().includes(q) ||
						(r.vendor || "").toLowerCase().includes(q)
				)
			: me.rows;

		if (!rows.length) {
			tbody.html(`<tr><td colspan="7" class="text-center text-muted">${__("No apps match the filters")}</td></tr>`);
			return;
		}

		tbody.html(
			rows
				.map((r) => {
					const delta = me.compute_delta(r);
					const deltaCell =
						delta == null
							? "—"
							: `<span style="color:${Math.abs(delta) > 25 ? "#FF9800" : "var(--text-muted)"};font-weight:500">${delta > 0 ? "+" : ""}${delta.toFixed(1)}%</span>`;
					const actual = r.actual_amount == null ? "" : r.actual_amount;
					return `
				<tr data-app="${frappe.utils.escape_html(r.app)}">
					<td><a href="/app/saas-application/${frappe.utils.escape_html(r.app)}">${frappe.utils.escape_html(r.app_name)}</a></td>
					<td><span class="badge" style="background:#E8F4FE;color:#2490EF;font-size:11px;font-weight:500;padding:2px 6px;border-radius:3px">${frappe.utils.escape_html(r.subscription_model || "")}</span></td>
					<td style="color:var(--text-muted)">${frappe.utils.escape_html(r.cost_center || "—")}</td>
					<td class="text-right" style="color:var(--text-muted);font-variant-numeric:tabular-nums">${me.fmt(r.expected_amount, r.currency)}</td>
					<td class="text-right" style="font-variant-numeric:tabular-nums;background:#FFFBEA">
						<input type="number" class="mce-actual form-control"
							style="text-align:right;background:transparent;border:none;padding:2px 4px;font-variant-numeric:tabular-nums"
							value="${actual}"
							placeholder="0">
					</td>
					<td class="text-right">${deltaCell}</td>
					<td><input type="text" class="mce-note form-control" style="border:none;padding:2px 4px;font-size:12px" value="${frappe.utils.escape_html(r.notes || "")}" placeholder="…"></td>
				</tr>`;
				})
				.join("")
		);

		// Bind input handlers
		tbody.find("input.mce-actual").on("blur", function () {
			const tr = $(this).closest("tr");
			const app = tr.data("app");
			const val = $(this).val();
			const note = tr.find("input.mce-note").val();
			me.upsert_cell(app, val, note);
		});

		tbody.find("input.mce-note").on("blur", function () {
			const tr = $(this).closest("tr");
			const app = tr.data("app");
			const val = tr.find("input.mce-actual").val();
			const note = $(this).val();
			if (val !== "") me.upsert_cell(app, val, note);
		});
	}

	compute_delta(r) {
		if (r.expected_amount == null || r.expected_amount === 0) return null;
		if (r.actual_amount == null) return null;
		return ((r.actual_amount - r.expected_amount) / r.expected_amount) * 100;
	}

	fmt(amount, currency) {
		if (amount == null || amount === "") return "—";
		return format_currency(amount, currency || "AED");
	}

	// ---------------------------------------------------------------------
	// Save
	// ---------------------------------------------------------------------

	async upsert_cell(app, amount, notes) {
		if (amount === "" || amount == null) return;
		try {
			const r = await frappe.call({
				method: "saas_register.saas_register.page.monthly_cost_entry.monthly_cost_entry.upsert_monthly_cost",
				args: { application: app, month: this.month, amount: parseFloat(amount), notes },
			});
			// Mirror saved value back into local cache for delta recompute
			const row = this.rows.find((r) => r.app === app);
			if (row) {
				row.actual_amount = parseFloat(amount);
				row.notes = notes || "";
			}
			frappe.show_alert({ message: __("Saved"), indicator: "green" }, 2);
			this.render_rows();
		} catch (e) {
			frappe.show_alert({ message: __("Save failed: {0}", [e.message || ""]), indicator: "red" });
		}
	}

	async save_all() {
		const me = this;
		const rows = me.page.main.find("tr[data-app]").toArray();
		let saved = 0;
		for (const tr of rows) {
			const $tr = $(tr);
			const app = $tr.data("app");
			const val = $tr.find("input.mce-actual").val();
			const note = $tr.find("input.mce-note").val();
			if (val === "" || val == null) continue;
			try {
				await frappe.call({
					method: "saas_register.saas_register.page.monthly_cost_entry.monthly_cost_entry.upsert_monthly_cost",
					args: { application: app, month: me.month, amount: parseFloat(val), notes: note },
				});
				saved++;
			} catch (e) {
				// keep going
			}
		}
		frappe.show_alert({ message: __("Saved {0} rows", [saved]), indicator: "green" });
		me.load();
	}

	// ---------------------------------------------------------------------
	// Paste dialog
	// ---------------------------------------------------------------------

	show_paste_dialog() {
		const me = this;
		const d = new frappe.ui.Dialog({
			title: __("Paste from spreadsheet"),
			fields: [
				{
					fieldname: "info",
					fieldtype: "HTML",
					options:
						"<p>Paste two tab- or comma-separated columns: <b>Application</b>, <b>Amount</b>. One row per line. App name matching is fuzzy.</p>",
				},
				{
					fieldname: "rows",
					fieldtype: "Code",
					label: __("Rows"),
					reqd: 1,
				},
			],
			primary_action_label: __("Preview"),
			primary_action: async (values) => {
				const lines = (values.rows || "").split(/\n/).map((l) => l.trim()).filter(Boolean);
				const parsed = lines.map((l) => {
					const parts = l.split(/\t|,/).map((p) => p.trim());
					return { app_name: parts[0], amount: parseFloat(parts[1]) };
				});
				const r = await frappe.call({
					method: "saas_register.saas_register.page.monthly_cost_entry.monthly_cost_entry.paste_monthly_costs",
					args: { month: me.month, rows: parsed, commit: 0 },
				});
				const { matched, unmatched } = r.message || { matched: [], unmatched: [] };
				const html = `
					<div style="font-size:12px">
						<p><b>${matched.length}</b> matched, <b>${unmatched.length}</b> unmatched.</p>
						${matched.length ? `<details><summary>Matched (${matched.length})</summary><ul>${matched.map((m) => `<li>${frappe.utils.escape_html(m.input)} → ${frappe.utils.escape_html(m.app)} (${frappe.utils.escape_html(m.app_name)}) → ${m.amount}</li>`).join("")}</ul></details>` : ""}
						${unmatched.length ? `<details open><summary>Unmatched (${unmatched.length})</summary><ul>${unmatched.map((u) => `<li>${frappe.utils.escape_html(u.app_name)} → ${u.amount}</li>`).join("")}</ul></details>` : ""}
					</div>`;
				d.fields_dict.info.$wrapper.html(html);
				d.set_primary_action(__("Confirm & save {0}", [matched.length]), async () => {
					await frappe.call({
						method: "saas_register.saas_register.page.monthly_cost_entry.monthly_cost_entry.paste_monthly_costs",
						args: { month: me.month, rows: parsed, commit: 1 },
					});
					frappe.show_alert({ message: __("Saved"), indicator: "green" });
					d.hide();
					me.load();
				});
			},
		});
		d.show();
	}
}
